// @nxg-org/mineflayer-util-plugin/lib/static/aabbUtil.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AABBUtils = void 0;
const aabb_1 = require("../calcs/aabb");
var AABBUtils;
(function (AABBUtils) {
    function getBlockAABB(block, height = 1) {
        const { x, y, z } = block.position;
        return new aabb_1.AABB(x, y, z, x + 1, y + height, z + 1);
    }
    AABBUtils.getBlockAABB = getBlockAABB;
    function getBlockPosAABB(block, height = 1) {
        const { x, y, z } = block.floored();
        return new aabb_1.AABB(x, y, z, x + 1, y + height, z + 1);
    }
    AABBUtils.getBlockPosAABB = getBlockPosAABB;
    function getEntityAABB(entity) {
        var _a;
        switch (entity.type) {
            case "player":
                return getPlayerAABB({ position: entity.position });
            case "mob":
            default: //TODO: Implement better AABBs. However, this may just be correct.
                return getEntityAABBRaw({ position: entity.position, height: entity.height, width: (_a = entity.width) !== null && _a !== void 0 ? _a : entity.height });
        }
    }
    AABBUtils.getEntityAABB = getEntityAABB;
    function getPlayerAABB(entity) {
        return getEntityAABBRaw({ position: entity.position, height: 1.8, width: 0.6 });
    }
    AABBUtils.getPlayerAABB = getPlayerAABB;
    function getEntityAABBRaw(entity) {
        const w = entity.width ? entity.width / 2 : entity.height / 2;
        const { x, y, z } = entity.position;
        return new aabb_1.AABB(-w, 0, -w, w, entity.height, w).offset(x, y, z);
    }
    AABBUtils.getEntityAABBRaw = getEntityAABBRaw;
})(AABBUtils = exports.AABBUtils || (exports.AABBUtils = {}));
